### DISCUSSION TITLE
Unable to pull package of public personal repository pushed to github package

### DISCUSSION THREAD

**Discussion by tasdikrahman** [community/community#26863] [Packages]
<p>Greetings,</p>
<p>Thanks for adding support for github packages to upload container images, I recently was trying out pushing packages to it, <a href="https://github.com/tasdikrahman/bhola/pull/55/checks?check_run_id=1140051698" rel="noopener nofollow ugc">https://github.com/tasdikrahman/bhola/pull/55/checks?check_run_id=1140051698</a>,</p>
<pre><code class="lang-auto">Successfully tagged docker.pkg.github.com/tasdikrahman/bhola/bhola:feat-issue-54
Pushing image [docker.pkg.github.com/tasdikrahman/bhola/bhola:feat-issue-54]
The push refers to repository [docker.pkg.github.com/tasdikrahman/bhola/bhola]
b03ea15ce214: Preparing
dbc12b000854: Preparing
0a7fec4e9e75: Preparing
07af6d27e52d: Preparing
43cd00501c28: Preparing
9c8e38fa209a: Preparing
708afaaf892f: Preparing
94928d5d49e7: Preparing
bcf2f368fe23: Preparing
9c8e38fa209a: Waiting
708afaaf892f: Waiting
94928d5d49e7: Waiting
bcf2f368fe23: Waiting
43cd00501c28: Layer already exists
9c8e38fa209a: Layer already exists
708afaaf892f: Layer already exists
94928d5d49e7: Layer already exists
bcf2f368fe23: Layer already exists
b03ea15ce214: Pushed
dbc12b000854: Pushed
0a7fec4e9e75: Pushed
07af6d27e52d: Pushed
feat-issue-54: digest: sha256:bc2c0c5f2e2eede136fc51b7de16a00781d787743c7cae1d2b500b7e46709cdb size: 2201
</code></pre>
<p>I tried pulling this image on my machine</p>
<pre><code class="lang-auto">$ docker pull docker.pkg.github.com/tasdikrahman/bhola/bhola:feat-issue-54
Error response from daemon: Get "https://docker.pkg.github.com/v2/tasdikrahman/bhola/bhola/manifests/feat-issue-54": no basic auth credentials
</code></pre>
<p>I tried checking the answer posted here <a href="https://github.community/t/just-published-a-new-gh-package-have-few-questions/132521/2" class="inline-onebox">Just published a new GH package. Have few questions</a>, but couldn’t find the setting to make my container packages public in case they are private.</p>
<p>Let me know if I am missing something folks.</p>

**Comment by tasdikrahman** [comment]
<p>Another thing which I noticed was that I can see the package under the public repositories</p>
<aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/tasdikrahman?tab=packages&amp;visibility=public" target="_blank" rel="noopener nofollow ugc">GitHub</a>
  </header>
  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/4672518/181100352-1ffc5f6c-90c2-4ac8-b125-c0c5ff606475.jpeg" class="thumbnail onebox-avatar" width="400" height="400">

<h3><a href="https://github.com/tasdikrahman?tab=packages&amp;visibility=public" target="_blank" rel="noopener nofollow ugc">tasdikrahman - Overview</a></h3>

<p>Engineering Platform @gojek, former SRE @razorpay. Weekend chef, Backpacker, past contributor to @oVirt (Redhat).
 - tasdikrahman</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by airtower-luna** [comment]
<p>I notice that your workflow pushes to the old GitHub Packages Docker registry (<code>docker.pkg.github.com</code>). If you want your packages to be publicly accessible without any authentication you need to use the new <a href="https://docs.github.com/en/packages/getting-started-with-github-container-registry/about-github-container-registry">GitHub Container Registry (<code>ghcr.io</code>)</a>. Note that new images are private by default, so after the first push you’d need to switch the image to public.</p>

**Comment by tasdikrahman** [comment]
<p>I get this error now</p>
<pre><code class="lang-auto"> Logging in to registry ghcr.io
WARNING! Using --password via the CLI is insecure. Use --password-stdin.
Error response from daemon: Get https://ghcr.io/v2/: denied
Error: exit status 1
exit status 1
Usage:
  github-actions build-push [flags]

Flags:
  -h, --help   help for build-push
</code></pre>
<p>The block for pushing to github package is this</p>
<pre><code class="lang-auto">      - name: Push to GitHub Packages
        uses: docker/build-push-action@v1
        with:
          username: tasdikrahman
          password: ${{ secrets.GITHUB_TOKEN }}
          registry: ghcr.io
          repository: tasdikrahman/bhola/bhola
          tag_with_ref: true
</code></pre>
<p>anything which I need to add/change.</p>

**Comment by airtower-luna** [comment]
<p>I see one thing that’s definitely a problem, and one that probably is:</p>
<ol>
<li>You can’t use GITHUB_TOKEN to authenticate with GHCR. You need to use a PAT. See <a href="https://docs.github.com/en/packages/getting-started-with-github-container-registry/migrating-to-github-container-registry-for-docker-images#authenticating-with-the-container-registry">Authenticating with the container registry</a>.</li>
<li>The <code>repository</code> parameter seems to follow the  <code>OWNER/REPOSITORY/IMAGE_NAME</code> pattern. The GHCR uses <code>OWNER/IMAGE_NAME</code>, see <a href="https://docs.github.com/en/packages/getting-started-with-github-container-registry/migrating-to-github-container-registry-for-docker-images#domain-changes">Domain changes</a>.</li>
</ol>
<p>To link your image to the source repository you can either set the <code>org.opencontainers.image.source</code> label to the repository path (that’s what I prefer), or manually link them using the web UI. See <a href="https://docs.github.com/en/packages/managing-container-images-with-github-container-registry/connecting-a-repository-to-a-container-image">Connecting a repository to a container image</a>.</p>

**Comment by jcansdale** [comment]
<aside class="quote no-group" data-username="airtower-luna" data-post="5" data-topic="133151">
<div class="title">
<div class="quote-controls"></div>
 airtower-luna:</div>
<blockquote>
<p>The <code>repository</code> parameter seems to follow the <code>OWNER/REPOSITORY/IMAGE_NAME</code> pattern. The GHCR uses <code>OWNER/IMAGE_NAME</code> , see <a href="https://docs.github.com/en/packages/getting-started-with-github-container-registry/migrating-to-github-container-registry-for-docker-images#domain-changes" rel="noopener nofollow ugc">Domain changes</a>.</p>
</blockquote>
</aside>
<p>You can actually use any number of components with <a href="http://ghcr.io" rel="noopener nofollow ugc">ghcr.io</a>. For example, <code>ghcr.io/OWNER/REPOSITORY/IMAGE_NAME</code> would also work. This might be useful if you have multiple images associated with the same repo.</p>
<p>At the moment the REPOSITORY component doesn’t automatically associate the image with a repository of the same name, but it hopefully will in a future update.</p>

**Comment by tasdikrahman** [comment]
<p>Thanks for the response, I am not using a PAT which has read, write and delete access for packages which I have added to my repo secrets.</p>
<p>this is what I have in the pipeline as of now</p>
<pre><code class="lang-auto">      - name: Check out the repo
        uses: actions/checkout@v2
      - name: Docker login
        run: &gt;
          echo ${{ secrets.GHCR_PAT }} | docker login ghcr.io -u tasdikrahman --password-stdin
      - name: Build container image
        run: &gt;
          docker build
          --label org.opencontainers.image.revision=${{ github.sha }}
          --iidfile bhola.id .
      - name: Show image information
        run: |
          docker image inspect $(cat bhola.id)
      - name: show docker images
        run: |
          docker images
      - name: take the first docker image from the list
        run: |
          ID=$(docker images -a | awk '{print $3}' | sed '1d' | head -1)
          echo $ID
      - name: Tag container image
        run: |
          docker tag ${ID} ghcr.io/tasdikrahman/bhola:${{ github.SHA }}
      - name: Push the container image
        run: |
          docker push ghcr.io/tasdikrahman/bhola:${{ github.SHA }}
      - name: Docker logout
        if: always()
        run: |
          docker logout ghcr.io
</code></pre>
<p>the failure now which I get is during the tagging part. Would fail, while I try to tag</p>
<pre><code class="lang-auto">Run docker tag ${ID} ghcr.io/***/bhola:df246a982a3385fc70d52792b1847db964a9f7a7
  docker tag ${ID} ghcr.io/***/bhola:df246a982a3385fc70d52792b1847db964a9f7a7
  shell: /bin/bash -e {0}
"docker tag" requires exactly 2 arguments.
See 'docker tag --help'.

Usage:  docker tag SOURCE_IMAGE[:TAG] TARGET_IMAGE[:TAG]

Create a tag TARGET_IMAGE that refers to SOURCE_IMAGE
##[error]Process completed with exit code 1.
</code></pre>
<p>I am not sure what am I doing wrong as I hardcoded the part of the OWNER here while it is still being picked up as <code>***/bhola:SHA</code></p>
<p>My pipeline for reference <a href="https://github.com/tasdikrahman/bhola/pull/55/checks?check_run_id=1144873993" rel="noopener nofollow ugc">https://github.com/tasdikrahman/bhola/pull/55/checks?check_run_id=1144873993</a></p>

**Comment by clarkbw** [comment, marked as answer]
<aside class="quote no-group" data-username="tasdikrahman" data-post="7" data-topic="133151">
<div class="title">
<div class="quote-controls"></div>
 tasdikrahman:</div>
<blockquote>
<pre><code class="lang-auto">        run: |
          ID=$(docker images -a | awk '{print $3}' | sed '1d' | head -1)
          echo $ID
      - name: Tag container image
        run: |
          docker tag ${ID} ghcr.io/tasdikrahman/bhola:${{ github.SHA }}
</code></pre>
</blockquote>
</aside>
<p>I think this is the issue. You can’t simply <code>echo</code> something and use it in the next run. Either you need to set it to output or an environment variable or combine it all into 1 run.</p>
<pre><code class="lang-auto">      - name: Tag the first docker image from the list
        run: |
          docker images
          ID=$(docker images -a | awk '{print $3}' | sed '1d' | head -1)
          echo $ID
          docker tag ${ID} ghcr.io/tasdikrahman/bhola:${{ github.SHA }}
</code></pre>

**Comment by tasdikrahman** [comment]
<aside class="quote no-group" data-username="airtower-luna" data-post="3" data-topic="133151">
<div class="title">
<div class="quote-controls"></div>
 airtower-luna:</div>
<blockquote>
<p>private</p>
</blockquote>
</aside>
<p>Awesome, that worked out folks!</p>
<p>Only thing was, the flow of making the image go from private to public by follow the instructions here <a href="https://docs.github.com/en/packages/managing-container-images-with-github-container-registry/connecting-a-repository-to-a-container-image" rel="noopener nofollow ugc">https://docs.github.com/en/packages/managing-container-images-with-github-container-registry/connecting-a-repository-to-a-container-image</a>, would it make sense to have a setting, where the container images of a repo are set to be public by default. What do you folks think</p>

**Comment by tasdikrahman** [comment]
<p>Also, would there be a container image retention policy similar to how docker hub is doing for free images?</p>

**Comment by clarkbw** [comment]
<aside class="quote no-group" data-username="tasdikrahman" data-post="9" data-topic="133151">
<div class="title">
<div class="quote-controls"></div>
 tasdikrahman:</div>
<blockquote>
<p>would it make sense to have a setting, where the container images of a repo are set to be public by default. What do you folks think</p>
</blockquote>
</aside>
<p>Good feedback, thank you!  We’ve been looking at ways to make this happen.</p>
